In [23]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np
from sklearn.ensemble import RandomForestRegressor
import pandas as pd


Imported all required libraries for data analysis, visualization, preprocessing, model building, and evaluation.

In [24]:
df = pd.read_csv("Housing.csv")
df.head()

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished
3,12215000,7500,4,2,2,yes,no,yes,no,yes,3,yes,furnished
4,11410000,7420,4,1,2,yes,yes,yes,no,yes,2,no,furnished


Loaded the housing dataset into a Pandas DataFrame for analysis.

In [25]:
df.shape

(545, 13)

Checked the number of rows and columns present in the dataset.

In [26]:
df.columns

Index(['price', 'area', 'bedrooms', 'bathrooms', 'stories', 'mainroad',
       'guestroom', 'basement', 'hotwaterheating', 'airconditioning',
       'parking', 'prefarea', 'furnishingstatus'],
      dtype='object')

Displayed all feature names available in the dataset.

In [27]:
df.isnull().sum()

price               0
area                0
bedrooms            0
bathrooms           0
stories             0
mainroad            0
guestroom           0
basement            0
hotwaterheating     0
airconditioning     0
parking             0
prefarea            0
furnishingstatus    0
dtype: int64

Verified that the dataset contains no missing values.

In [28]:
df.duplicated().sum()

np.int64(0)

No duplicate records were found in the dataset.

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


Examined data types, column information, and non-null values.

In [30]:
yes_no_cols = ['mainroad', 'guestroom', 'basement', 'hotwaterheating', 'airconditioning', 'prefarea']

for col in yes_no_cols:
    df[col] = df[col].map({'yes': 1, 'no': 0})


df['furnishingstatus'] = df['furnishingstatus'].map({'unfurnished': 0, 'semi-furnished': 1, 'furnished': 2})
df.head()

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,1,0,0,0,1,2,1,2
1,12250000,8960,4,4,4,1,0,0,0,1,3,0,2
2,12250000,9960,3,2,2,1,0,1,0,0,2,1,1
3,12215000,7500,4,2,2,1,0,1,0,1,3,1,2
4,11410000,7420,4,1,2,1,1,1,0,1,2,0,2


Converted categorical variables into numerical values suitable for machine learning models.

In [31]:
X = df.drop('price', axis=1)
y = df['price']

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

Separated the dataset into training and testing sets for model evaluation.

In [32]:
scaler=StandardScaler()
X_train_scaled=scaler.fit_transform(X_train)
X_test_scaled=scaler.transform(X_test)

Standardized feature values to bring them onto a similar scale.

In [33]:
X_train_scaled

array([[ 0.38416819,  0.05527092,  1.53917323, ...,  0.36795665,
        -0.55262032,  1.40940148],
       [ 0.9291807 ,  0.05527092,  1.53917323, ...,  2.70998729,
        -0.55262032,  0.07931697],
       [-0.60775457, -1.28351359, -0.5579503 , ...,  1.53897197,
        -0.55262032,  1.40940148],
       ...,
       [-0.29709744,  0.05527092,  1.53917323, ...,  0.36795665,
        -0.55262032,  1.40940148],
       [-0.5060189 , -1.28351359, -0.5579503 , ..., -0.80305867,
        -0.55262032, -1.25076755],
       [ 0.15707965,  0.05527092,  1.53917323, ...,  0.36795665,
        -0.55262032,  0.07931697]], shape=(436, 12))

Training data was successfully standardized using StandardScaler.

In [34]:
lr = LinearRegression()

lr.fit(X_train, y_train)

lr_pred = lr.predict(X_test)

Trained a Linear Regression model to predict house prices.

In [35]:
lr_mae = mean_absolute_error(y_test, lr_pred)
lr_rmse = np.sqrt(mean_squared_error(y_test, lr_pred))
lr_r2 = r2_score(y_test, lr_pred)

print("Linear Regression Results")
print("MAE :", lr_mae)
print("RMSE:", lr_rmse)
print("R²  :", lr_r2)

Linear Regression Results
MAE : 979679.6912959905
RMSE: 1331071.4167895105
R²  : 0.6494754192267804


Evaluated the Linear Regression model using MAE, RMSE, and R² Score.

In [36]:
rf = RandomForestRegressor(
    n_estimators=200,
    random_state=42
)

rf.fit(X_train, y_train)

rf_pred = rf.predict(X_test)

Trained a Random Forest Regressor to compare performance with Linear Regression.

In [37]:
rf_mae = mean_absolute_error(y_test, rf_pred)
rf_rmse = np.sqrt(mean_squared_error(y_test, rf_pred))
rf_r2 = r2_score(y_test, rf_pred)

print("Random Forest Results")
print("MAE :", rf_mae)
print("RMSE:", rf_rmse)
print("R²  :", rf_r2)

Random Forest Results
MAE : 1011501.4606269113
RMSE: 1388576.3746170169
R²  : 0.6185344640262593


Evaluated the Random Forest model using MAE, RMSE, and R² Score.

In [38]:
comparison = pd.DataFrame({
    'Model': ['Linear Regression', 'Random Forest'],
    'MAE': [lr_mae, rf_mae],
    'RMSE': [lr_rmse, rf_rmse],
    'R2 Score': [lr_r2, rf_r2]
})

comparison

,Model,MAE,RMSE,R2 Score
0,Linear Regression,9.796797e+05,1.331071e+06,0.649475
1,Random Forest,1.011501e+06,1.388576e+06,0.618534


Compared the performance of both models based on error metrics and R² Score.

In [39]:
importance = pd.DataFrame({
    "Feature": X.columns,
    "Importance": rf.feature_importances_
})

importance = importance.sort_values(
    by="Importance",
    ascending=False
)

importance.head(10)

,Feature,Importance
0,area,0.468806
2,bathrooms,0.153014
8,airconditioning,0.060200
9,parking,0.056942
3,stories,0.056363
11,furnishingstatus,0.051359
1,bedrooms,0.046711
6,basement,0.033100
10,prefarea,0.028503
7,hotwaterheating,0.017928


Identified the most influential features affecting house prices.

In [40]:
sns.set_style("whitegrid")
df = pd.read_csv("Housing.csv")  # update path as needed

# Encode binary yes/no columns for correlation purposes
binary_cols = ["mainroad", "guestroom", "basement", "hotwaterheating",
               "airconditioning", "prefarea"]
df_enc = df.copy()
for col in binary_cols:
    df_enc[col] = df_enc[col].map({"yes": 1, "no": 0})

furnishing_map = {"unfurnished": 0, "semi-furnished": 1, "furnished": 2}
df_enc["furnishingstatus"] = df_enc["furnishingstatus"].map(furnishing_map)


In [41]:
# ---------- Chart 1: Histogram of house prices ----------
plt.figure(figsize=(9, 6))
plt.hist(df["price"], bins=30, color="#4C72B0", edgecolor="white")
plt.title("Distribution of House Prices", fontsize=14, fontweight="bold")
plt.xlabel("Price (₹)")
plt.ylabel("Number of Houses")
plt.gca().xaxis.set_major_formatter(plt.FuncFormatter(lambda x, _: f"{x/1e6:.0f}M"))
plt.tight_layout()
plt.savefig("chart1_price_histogram.png", dpi=150)
plt.close()

This chart shows the distribution of house prices in the dataset.

In [42]:
# ---------- Chart 2: Correlation heatmap ----------
corr = df_enc.corr(numeric_only=True)
plt.figure(figsize=(10, 8))
sns.heatmap(corr, annot=True, fmt=".2f", cmap="coolwarm", center=0,
            square=True, linewidths=0.5, cbar_kws={"shrink": 0.8})
plt.title("Correlation Heatmap of House Features", fontsize=14, fontweight="bold")
plt.tight_layout()
plt.savefig("chart2_correlation_heatmap.png", dpi=150)
plt.close()

This chart shows the correlation between different housing features and house prices.

In [43]:
# ---------- Chart 3: Price vs Area scatter, colored by furnishing status ----------
plt.figure(figsize=(9, 6))
colors = {"unfurnished": "#DD8452", "semi-furnished": "#55A868", "furnished": "#4C72B0"}
for status, group in df.groupby("furnishingstatus"):
    plt.scatter(group["area"], group["price"], label=status,
                color=colors[status], alpha=0.6, edgecolor="white", s=50)

plt.title("Price vs. Area by Furnishing Status", fontsize=14, fontweight="bold")
plt.xlabel("Area (sq ft)")
plt.ylabel("Price (₹)")
plt.gca().yaxis.set_major_formatter(plt.FuncFormatter(lambda y, _: f"{y/1e6:.0f}M"))
plt.legend(title="Furnishing Status")
plt.tight_layout()
plt.savefig("chart3_price_vs_area.png", dpi=150)
plt.close()

print("Charts saved.")
print(corr["price"].sort_values(ascending=False))

Charts saved.
price               1.000000
area                0.535997
bathrooms           0.517545
airconditioning     0.452954
stories             0.420712
parking             0.384394
bedrooms            0.366494
prefarea            0.329777
furnishingstatus    0.304721
mainroad            0.296898
guestroom           0.255517
basement            0.187057
hotwaterheating     0.093073
Name: price, dtype: float64


This chart shows the relationship between house area and price across different furnishing categories.

In [44]:
# ---------- Quick model for accuracy reference ----------
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_absolute_error

X = df_enc.drop(columns=["price"])
y = df_enc["price"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
model = LinearRegression().fit(X_train, y_train)
pred = model.predict(X_test)
r2 = r2_score(y_test, pred)
mae = mean_absolute_error(y_test, pred)
mape = (abs((y_test - pred) / y_test)).mean() * 100
print(f"\nModel R2: {r2:.3f} | MAE: {mae:,.0f} | MAPE: {mape:.1f}%")



Model R2: 0.649 | MAE: 979,680 | MAPE: 21.3%


Insights & SUMMARY
------------------
Area, bathrooms, and air conditioning turned out to be the strongest drivers
of house price, with area and bathroom count showing the highest correlations
(0.52-0.54) - bigger, more bathroom-rich homes consistently command higher
prices. A simple linear regression model explained about 65% of the price
variation (R2=0.65), with predictions typically off by around 980,000,
or roughly 21% in either direction - decent for a first pass, but not precise
enough to price a property with confidence. What stood out was how weak hot
water heating's relationship to price was approximately (0.09) despite seeming like a
meaningful amenity, while preferred-area location mattered more than expected
for a binary feature. For a real estate business, this suggests marketing and
pricing strategy should lean heavily on square footage and bathroom count as
the key value levers, while features like hot water heating add little resale
value and shouldn't be over-invested in for renovations aimed at boosting price.